samza-kv-couchbase/src/main/java/org/apache/samza/table/remote/couchbase/CouchbaseBucketRegistry.java

```diff
@@ -0,0 +1,181 @@
+   * bucket name. Each time this method is called, the counter of usage of the bucket is increased by one.
+   * @param bucketName name of the bucket to be opened
+   * @param clusterNodes list of cluster nodes
+   * @param configs couchbase environment configs
+   * @return A Bucket instance associated with the bucket name and cluster nodes
+   */
+  public synchronized Bucket getBucket(String bucketName, List<String> clusterNodes,
+      CouchbaseEnvironmentConfigs configs) {
+    String bucketId = getBucketId(bucketName, clusterNodes);
+    String clusterId = getClusterId(clusterNodes);
+    if (!openedClusters.containsKey(clusterId)) {
+      openedClusters.put(clusterId, openCluster(clusterNodes, configs));
+    }
+    if (!openedBuckets.containsKey(bucketId)) {
+      openedBuckets.put(bucketId, openBucket(bucketName, openedClusters.get(clusterId)));
+    }
+    bucketUsageCounts.put(bucketId, bucketUsageCounts.getOrDefault(bucketId, 0) + 1);
+    clusterUsageCounts.put(clusterId, clusterUsageCounts.getOrDefault(clusterId, 0) + 1);
+    return openedBuckets.get(bucketId);
+  }
+
+  /**
+   * A synchronized method to close a bucket given the bucket name. Each time this method is called, the counter of
+   * usage of the bucket is decreased by one. Only when the counter becomes zero will the bucket be actually closed.
+   * Same for cluster, only when all buckets within a cluster have been closed will the cluster be closed.
+   * @param bucketName name of the bucket to be opened
+   * @param clusterNodes list of cluster nodes
+   * @return true if bucket is closed successfully, otherwise false.
+   */
+  public synchronized boolean closeBucket(String bucketName, List<String> clusterNodes) {
+    String bucketId = getBucketId(bucketName, clusterNodes);
+    String clusterId = getClusterId(clusterNodes);
+    if (!openedBuckets.containsKey(bucketId) || !openedClusters.containsKey(clusterId)) {
+      return false;
+    }
+    bucketUsageCounts.put(bucketId, bucketUsageCounts.get(bucketId) - 1);
+    clusterUsageCounts.put(clusterId, clusterUsageCounts.get(clusterId) - 1);
+    Boolean bucketClosed = true;
+    Boolean clusterClosed = true;
+    if (bucketUsageCounts.get(bucketId) == 0) {
+      bucketClosed = openedBuckets.get(bucketId).close();
+      openedBuckets.remove(bucketId);
+      bucketUsageCounts.remove(bucketId);
+      if (clusterUsageCounts.get(clusterId) == 0) {
+        clusterClosed = openedClusters.get(clusterId).disconnect();
+        openedClusters.remove(clusterId);
+        clusterUsageCounts.remove(clusterId);
+      }
+    }
+    return bucketClosed && clusterClosed;
+  }
+
+  /**
+   * Helper method to open a cluster given cluster nodes and environment configurations.
+   */
+  private Cluster openCluster(List<String> clusterNodes, CouchbaseEnvironmentConfigs configs) {
+    DefaultCouchbaseEnvironment.Builder envBuilder = new DefaultCouchbaseEnvironment.Builder();
+    if (configs.sslEnabled != null) {
```

Comment:
do you need all the null check?

Reply:
I think so because we don't have default values for the environment configs. We can't provide default values because it depends on use cases. So it's better to let user set it when they need to.